src/utils/recommendations/helper/vectorBasedMapping.js

```diff
@@ -0,0 +1,50 @@
+/**
+ * Calculate the cosine similarity between two vectors.
+ * @param {Object.<string, number>} queryVector - The query vector.
+ * @param {Object.<string, number>} documentVector - The document vector.
+ * @returns {number} The cosine similarity between the vectors./
+ */
+
+function calculateCosineSimilarity(queryVector, documentVector) {
+  // Early return for empty vectors
+  if (Object.keys(queryVector).length === 0 || Object.keys(documentVector).length === 0) {
+    return 0;
+  }
+
+  let dotProduct = 0;
+  let queryVectorModulusSquared = 0;
+  let documentVectorModulusSquared = 0;
```

Comment:
Please rename these two variables to queryVectorSquared & docuemntVectorSquared since the modulus of a vector is when we take the square root of this square